pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -101,12 +101,41 @@ func (r Requirements) Add(requirements ...v1.NodeSelectorRequirement) Requiremen
 		case v1.NodeSelectorOpDoesNotExist:
 			values = sets.NewSet()
 		}
-		if existing, ok := r.requirements[requirement.Key]; ok {
+		if existing, ok := cp.requirements[requirement.Key]; ok {
 			values = values.Intersection(existing)
 		}
-		r.requirements[requirement.Key] = values
+		cp.requirements[requirement.Key] = values
+	}
+	cp.rebuild()
+	return cp
+}
+
+// rebuild re-generates the node selector requirements based on the set based versions.  This improves scheduling speed
+// as it causes us to not carry around redundant requirements (e.g. 20x copies of instance-type in [it0, it1, ..., it400])
+func (r *Requirements) rebuild() {
```

Comment:
I think we should refactor requirements in general. Perhaps worth punting on this until then. Happy to leave this in, but it bloats an already massive PR.

Reply:
It greatly affects performance, I'd like to leave it in this one for now, or possibly just cherry-pick it to main to reduce this PRs size.  Need to get the benchmark CI in as it's very easy to make a change and send scheduling performance through the floor.